src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -423,33 +423,49 @@ private ImmutableArray<SourceMethodTypeParameterSymbol> MakeTypeParameters(Diagn
             return result.ToImmutableAndFree();
         }
 
-        public override ImmutableArray<TypeParameterConstraintClause> TypeParameterConstraintClauses
+        public override ImmutableArray<TypeParameterConstraintClause> GetTypeParameterConstraintClauses(bool early)
         {
-            get
+            var clauses = _lazyTypeParameterConstraints;
+            if (clauses.IsDefault)
             {
-                if (_lazyTypeParameterConstraints.IsDefault)
+                // Early step.
+                var diagnostics = DiagnosticBag.GetInstance();
+                var constraints = this.MakeTypeParameterConstraintsEarly(
+                    _binder,
+                    TypeParameters,
+                    _syntax.ConstraintClauses,
+                    _syntax.Identifier.GetLocation(),
+                    diagnostics);
+                lock (_declarationDiagnostics)
```

Comment:
Consider another check to `_lazyTypeParameterConstraints.IsDefault`, as `MakeTypeParameterConstraintsEarly` is not a small function.

Reply:
If `_lazyTypeParameterConstraints.IsDefault` fails, another thread has already updated the field so there shouldn't be any lock contention from with that thread.

---
In reply to: [198667734](https://github.com/dotnet/roslyn/pull/27525#discussion_r198667734) [](ancestors = 198667734)